File: crates/better-core/src/ai/insights.rs
// crates/better-core/src/ai/insights.rs
// Cross-Project Intelligence — org-level dependency analysis

use serde::Serialize;
use std::collections::HashMap;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, Serialize)]
pub struct OrgInsights {
    pub projects_analyzed: usize,
    pub total_unique_deps: usize,
    pub total_dep_instances: usize,
    pub version_inconsistencies: Vec<VersionInconsistency>,
    pub consolidation_opportunities: Vec<ConsolidationOpportunity>,
    pub standardization_score: u8,
}

#[derive(Debug, Clone, Serialize)]
pub struct VersionInconsistency {
    pub package: String,
    pub versions: HashMap<String, Vec<String>>,  // version -> [project names]
    pub recommended: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct ConsolidationOpportunity {
    pub category: String,
    pub packages: Vec<String>,
    pub projects: Vec<String>,
    pub recommended: String,
    pub reason: String,
}

#[derive(Debug, Clone)]
struct ProjectDeps {
    name: String,
    path: PathBuf,
    deps: HashMap<String, String>,  // name -> version
}

// Known consolidation patterns
const CONSOLIDATIONS: &[(&str, &[&str], &str)] = &[
    ("http-client", &["axios", "node-fetch", "got", "superagent", "ky"], "Pick one HTTP client org-wide"),
    ("date-library", &["moment", "date-fns", "dayjs", "luxon"], "Use date-fns or dayjs everywhere"),
    ("test-framework", &["jest", "mocha", "vitest", "jasmine"], "Standardize on one test framework"),
    ("bundler", &["webpack", "vite", "rollup", "parcel", "esbuild"], "Standardize on one bundler"),
    ("logger", &["winston", "pino", "bunyan", "log4js"], "Standardize on one logger"),
    ("orm", &["sequelize", "typeorm", "prisma", "drizzle-orm", "mongoose"], "Standardize on one ORM"),
];

/// Analyze dependencies across all projects in a directory.
pub fn analyze_org(root_dir: &Path) -> Result<OrgInsights, String> {
    let projects = discover_projects(root_dir)?;
    if projects.is_empty() {
        return Ok(OrgInsights {
            projects_analyzed: 0,
            total_unique_deps: 0,
            total_dep_instances: 0,
            version_inconsistencies: vec![],
            consolidation_opportunities: vec![],
            standardization_score: 100,
        });
    }

    let mut all_deps: HashMap<String, HashMap<String, Vec<String>>> = HashMap::new();
    let mut total_instances = 0;

    for project in &projects {
        for (dep, version) in &project.deps {
            total_instances += 1;
            all_deps
                .entry(dep.clone())
                .or_default()
                .entry(version.clone())
                .or_default()
                .push(project.name.clone());
        }
    }

    // Find version inconsistencies (same package, different versions across projects)
    let mut inconsistencies = vec![];
    for (pkg, versions) in &all_deps {
        if versions.len() > 1 {
            // Find the most common version as recommendation
            let recommended = versions.iter()
                .max_by_key(|(_, projects)| projects.len())
                .map(|(v, _)| v.clone())
                .unwrap_or_default();

            inconsistencies.push(VersionInconsistency {
                package: pkg.clone(),
                versions: versions.clone(),
                recommended,
            });
        }
    }
    inconsistencies.sort_by(|a, b| b.versions.len().cmp(&a.versions.len()));

    // Find consolidation opportunities
    let mut consolidations = vec![];
    for (category, packages, reason) in CONSOLIDATIONS {
        let found: Vec<(&str, Vec<String>)> = packages.iter()
            .filter_map(|&pkg| {
                all_deps.get(pkg).map(|versions| {
                    let projects: Vec<String> = versions.values().flatten().cloned().collect();
                    (pkg, projects)
                })
            })
            .collect();

        if found.len() > 1 {
            let all_projects: Vec<String> = found.iter()
                .flat_map(|(_, ps)| ps.iter().cloned())
                .collect::<std::collections::HashSet<_>>()
                .into_iter()
                .collect();
            let pkg_names: Vec<String> = found.iter().map(|(p, _)| p.to_string()).collect();

            consolidations.push(ConsolidationOpportunity {
                category: category.to_string(),
                packages: pkg_names,
                projects: all_projects,
                recommended: found.iter().max_by_key(|(_, ps)| ps.len())
                    .map(|(p, _)| p.to_string())
                    .unwrap_or_default(),
                reason: reason.to_string(),
            });
        }
    }

    // Standardization score: penalize per inconsistency and per consolidation opportunity
    let score = (100i32
        - inconsistencies.len() as i32 * 5
        - consolidations.len() as i32 * 10)
        .max(0) as u8;

    Ok(OrgInsights {
        projects_analyzed: projects.len(),
        total_unique_deps: all_deps.len(),
        total_dep_instances: total_instances,
        version_inconsistencies: inconsistencies,
        consolidation_opportunities: consolidations,
        standardization_score: score,
    })
}

fn discover_projects(root: &Path) -> Result<Vec<ProjectDeps>, String> {
    let mut projects = vec![];

    let entries = std::fs::read_dir(root)
        .map_err(|e| format!("Cannot read directory: {}", e))?;

    for entry in entries.flatten() {
        let path = entry.path();
        if !path.is_dir() { continue; }
        let name = path.file_name().unwrap_or_default().to_string_lossy().to_string();
        if name.starts_with('.') || name == "node_modules" { continue; }

        let pkg_path = path.join("package.json");
        if pkg_path.exists() {
            if let Some(deps) = read_npm_deps(&pkg_path) {
                projects.push(ProjectDeps { name, path, deps });
            }
        }
    }

    Ok(projects)
}

fn read_npm_deps(pkg_path: &Path) -> Option<HashMap<String, String>> {
    let content = std::fs::read_to_string(pkg_path).ok()?;
    let pkg: serde_json::Value = serde_json::from_str(&content).ok()?;
    let mut deps = HashMap::new();
    for key in &["dependencies", "devDependencies"] {
        if let Some(obj) = pkg.get(key).and_then(|v| v.as_object()) {
            for (k, v) in obj {
                if let Some(ver) = v.as_str() {
                    deps.insert(k.clone(), ver.to_string());
                }
            }
        }
    }
    Some(deps)
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;

    fn write_pkg(dir: &std::path::Path, content: &str) {
        std::fs::create_dir_all(dir).unwrap();
        let mut f = std::fs::File::create(dir.join("package.json")).unwrap();
        f.write_all(content.as_bytes()).unwrap();
    }

    #[test]
    fn empty_dir_returns_zero_projects() {
        let tmp = std::env::temp_dir().join("insights-test-empty");
        std::fs::create_dir_all(&tmp).unwrap();
        let insights = analyze_org(&tmp).unwrap();
        assert_eq!(insights.projects_analyzed, 0);
        assert_eq!(insights.standardization_score, 100);
        let _ = std::fs::remove_dir_all(&tmp);
    }

    #[test]
    fn single_project_no_inconsistencies() {
        let tmp = std::env::temp_dir().join("insights-test-single");
        write_pkg(&tmp, r#"{"name":"app","version":"1.0.0","dependencies":{"lodash":"^4.17.21"}}"#);
        let insights = analyze_org(&tmp).unwrap();
        // Single project → no inconsistencies by definition
        assert!(insights.version_inconsistencies.is_empty());
        let _ = std::fs::remove_dir_all(&tmp);
    }

    #[test]
    fn version_inconsistency_detected_across_projects() {
        let tmp = std::env::temp_dir().join("insights-test-inconsistent");
        write_pkg(&tmp.join("app-a"), r#"{"name":"app-a","version":"1.0.0","dependencies":{"lodash":"4.17.20"}}"#);
        write_pkg(&tmp.join("app-b"), r#"{"name":"app-b","version":"1.0.0","dependencies":{"lodash":"4.17.21"}}"#);
        let insights = analyze_org(&tmp).unwrap();
        assert!(!insights.version_inconsistencies.is_empty());
        assert!(insights.version_inconsistencies.iter().any(|i| i.package == "lodash"));
        let _ = std::fs::remove_dir_all(&tmp);
    }

    #[test]
    fn consolidation_opportunity_detected() {
        let tmp = std::env::temp_dir().join("insights-test-consolidate");
        write_pkg(&tmp.join("app-a"), r#"{"name":"app-a","version":"1.0.0","dependencies":{"moment":"^2.0.0"}}"#);
        write_pkg(&tmp.join("app-b"), r#"{"name":"app-b","version":"1.0.0","dependencies":{"dayjs":"^1.0.0"}}"#);
        let insights = analyze_org(&tmp).unwrap();
        assert!(!insights.consolidation_opportunities.is_empty());
        let _ = std::fs::remove_dir_all(&tmp);
    }

    #[test]
    fn consolidations_constant_is_nonempty() {
        assert!(!CONSOLIDATIONS.is_empty());
        assert!(CONSOLIDATIONS.iter().any(|(cat, _, _)| *cat == "http-client"));
    }

    #[test]
    fn analyze_org_same_version_no_inconsistencies() {
        let tmp = std::env::temp_dir().join("insights-test-same-ver");
        // Two projects, same dep, same version → no inconsistency
        write_pkg(&tmp.join("p1"), r#"{"name":"p1","version":"1.0.0","dependencies":{"lodash":"4.17.21"}}"#);
        write_pkg(&tmp.join("p2"), r#"{"name":"p2","version":"1.0.0","dependencies":{"lodash":"4.17.21"}}"#);
        let insights = analyze_org(&tmp).unwrap();
        assert!(insights.version_inconsistencies.is_empty());
        assert_eq!(insights.projects_analyzed, 2);
        let _ = std::fs::remove_dir_all(&tmp);
    }

    #[test]
    fn standardization_score_decreases_with_inconsistencies() {
        let tmp = std::env::temp_dir().join("insights-test-score");
        // 3 projects with lodash version differences → score < 100
        write_pkg(&tmp.join("pa"), r#"{"name":"pa","dependencies":{"lodash":"4.17.20"}}"#);
        write_pkg(&tmp.join("pb"), r#"{"name":"pb","dependencies":{"lodash":"4.17.21"}}"#);
        write_pkg(&tmp.join("pc"), r#"{"name":"pc","dependencies":{"lodash":"4.16.0"}}"#);
        let insights = analyze_org(&tmp).unwrap();
        assert!(insights.standardization_score < 100);
        let _ = std::fs::remove_dir_all(&tmp);
    }

    #[test]
    fn analyze_org_counts_total_instances() {
        let tmp = std::env::temp_dir().join("insights-test-instances");
        write_pkg(&tmp.join("q1"), r#"{"name":"q1","dependencies":{"a":"1.0","b":"2.0"}}"#);
        write_pkg(&tmp.join("q2"), r#"{"name":"q2","dependencies":{"a":"1.0","c":"3.0"}}"#);
        let insights = analyze_org(&tmp).unwrap();
        // q1 has 2 deps, q2 has 2 deps → 4 total instances
        assert_eq!(insights.total_dep_instances, 4);
        // Unique: a, b, c → 3
        assert_eq!(insights.total_unique_deps, 3);
        let _ = std::fs::remove_dir_all(&tmp);
    }
}

File: src/commands/insights.js
import { parseArgs } from "node:util";
import path from "node:path";
import { printJson, printText } from "../lib/output.js";
import { getRuntimeConfig } from "../lib/config.js";
import { runAnalyzeOrgNapi } from "../lib/core.js";
import fs from "node:fs/promises";
import { join } from "node:path";

// Known consolidation patterns
const CONSOLIDATIONS = [
  { category: "http-client", packages: ["axios", "node-fetch", "got", "superagent", "ky"], recommendation: "Pick one HTTP client org-wide" },
  { category: "date-library", packages: ["moment", "date-fns", "dayjs", "luxon"], recommendation: "Use date-fns or dayjs everywhere" },
  { category: "test-framework", packages: ["jest", "mocha", "vitest", "jasmine"], recommendation: "Standardize on one test framework" },
  { category: "bundler", packages: ["webpack", "vite", "rollup", "parcel", "esbuild"], recommendation: "Standardize on one bundler" },
  { category: "logger", packages: ["winston", "pino", "bunyan", "log4js"], recommendation: "Standardize on one logger" },
  { category: "orm", packages: ["sequelize", "typeorm", "prisma", "drizzle-orm", "mongoose"], recommendation: "Standardize on one ORM" },
];

/**
 * `better insights [DIR]` — org-level cross-project dependency analysis
 */
export async function cmdInsights(argv) {
  if (argv.includes("--help") || argv.includes("-h")) {
    printText(`Usage: better insights [DIR] [options]

Org-level cross-project dependency analysis.
Analyzes all subdirectories with package.json files.

Options:
  --json       Machine-readable output
  -h, --help   Show this help

Examples:
  better insights           # analyze subdirs of current directory
  better insights ~/projects
`);
    return;
  }

  const runtime = getRuntimeConfig();
  const { values, positionals } = parseArgs({
    args: argv,
    options: {
      json: { type: "boolean", default: runtime.json === true },
    },
    allowPositionals: true,
    strict: false,
  });

  const useJson = values.json;
  const rootDir = positionals[0] ? path.resolve(positionals[0]) : process.cwd();

  // NAPI fast path: Rust analyze_org
  const napiResult = runAnalyzeOrgNapi(rootDir);
  if (napiResult?.ok && napiResult.data && napiResult.data.projects_analyzed > 0) {
    const r = napiResult.data;
    const result = {
      ok: true,
      kind: "better.insights",
      projects_analyzed: r.projects_analyzed,
      total_unique_deps: r.total_unique_deps,
      total_dep_instances: r.total_dep_instances,
      version_inconsistencies: r.version_inconsistencies ?? [],
      consolidation_opportunities: r.consolidation_opportunities ?? [],
      standardization_score: r.standardization_score,
    };
    if (useJson) { printJson(result); }
    else {
      printText(`\nbetter — Org Insights (${r.projects_analyzed} projects)\n`);
      printText(`Unique dependencies: ${r.total_unique_deps} (${r.total_dep_instances} total instances)`);
      printText(`Standardization score: ${r.standardization_score}/100\n`);
      if (r.version_inconsistencies?.length > 0) {
        printText(`Version inconsistencies (${r.version_inconsistencies.length}):`);
        for (const inc of r.version_inconsistencies.slice(0, 10)) {
          const versions = Object.keys(inc.versions).join(", ");
          printText(`  ${inc.package}: ${versions} → recommend ${inc.recommended}`);
        }
      }
      if (r.consolidation_opportunities?.length > 0) {
        printText(`\nConsolidation opportunities:`);
        for (const opp of r.consolidation_opportunities) {
          printText(`  ${opp.category}: ${opp.packages.join(" + ")} → ${opp.recommended}`);
          printText(`    ${opp.reason} (affects ${opp.projects.length} projects)`);
        }
      }
      if (!r.version_inconsistencies?.length && !r.consolidation_opportunities?.length) {
        printText("Dependencies are well-standardized across projects.");
      }
    }
    return;
  }

  // JS fallback: Discover projects
  const projects = [];
  try {
    const entries = await fs.readdir(rootDir, { withFileTypes: true });
    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      if (entry.name.startsWith(".") || entry.name === "node_modules") continue;
      const pkgPath = join(rootDir, entry.name, "package.json");
      try {
        const pkg = JSON.parse(await fs.readFile(pkgPath, "utf8"));
        const deps = { ...pkg.dependencies, ...pkg.devDependencies };
        projects.push({ name: entry.name, path: join(rootDir, entry.name), deps });
      } catch {}
    }
  } catch (err) {
    const msg = `Cannot read directory: ${err.message}`;
    if (useJson) { printJson({ ok: false, error: msg }); } else { printText(`Error: ${msg}`); }
    process.exitCode = 1;
    return;
  }

  if (projects.length === 0) {
    if (useJson) { printJson({ ok: true, kind: "better.insights", projects_analyzed: 0, message: "No projects found" }); }
    else { printText("No projects with package.json found in subdirectories."); }
    return;
  }

  // Aggregate all dependencies
  const allDeps = new Map(); // pkg -> Map(version -> [projects])
  let totalInstances = 0;
  for (const project of projects) {
    for (const [dep, version] of Object.entries(project.deps || {})) {
      totalInstances++;
      if (!allDeps.has(dep)) allDeps.set(dep, new Map());
      const versions = allDeps.get(dep);
      if (!versions.has(version)) versions.set(version, []);
      versions.get(version).push(project.name);
    }
  }

  // Find version inconsistencies
  const inconsistencies = [];
  for (const [pkg, versions] of allDeps) {
    if (versions.size > 1) {
      const recommended = [...versions.entries()].sort((a, b) => b[1].length - a[1].length)[0][0];
      inconsistencies.push({ package: pkg, versions: Object.fromEntries(versions), recommended });
    }
  }
  inconsistencies.sort((a, b) => Object.keys(b.versions).length - Object.keys(a.versions).length);

  // Find consolidation opportunities
  const consolidationOpps = [];
  for (const { category, packages, recommendation } of CONSOLIDATIONS) {
    const found = packages.filter(p => allDeps.has(p));
    if (found.length > 1) {
      const projectsAffected = new Set();
      for (const pkg of found) {
        for (const projs of allDeps.get(pkg).values()) {
          for (const p of projs) projectsAffected.add(p);
        }
      }
      const mostUsed = found.reduce((a, b) => {
        const aCount = [...(allDeps.get(a)?.values() || [])].flat().length;
        const bCount = [...(allDeps.get(b)?.values() || [])].flat().length;
        return aCount >= bCount ? a : b;
      });
      consolidationOpps.push({ category, packages: found, projects: [...projectsAffected], recommended: mostUsed, reason: recommendation });
    }
  }

  const score = Math.max(0, 100 - inconsistencies.length * 5 - consolidationOpps.length * 10);

  const result = {
    ok: true,
    kind: "better.insights",
    projects_analyzed: projects.length,
    total_unique_deps: allDeps.size,
    total_dep_instances: totalInstances,
    version_inconsistencies: inconsistencies.slice(0, 20),
    consolidation_opportunities: consolidationOpps,
    standardization_score: score,
  };

  if (useJson) {
    printJson(result);
  } else {
    printText(`\nbetter — Org Insights (${projects.length} projects)\n`);
    printText(`Unique dependencies: ${allDeps.size} (${totalInstances} total instances)`);
    printText(`Standardization score: ${score}/100\n`);

    if (inconsistencies.length > 0) {
      printText(`Version inconsistencies (${inconsistencies.length}):`);
      for (const inc of inconsistencies.slice(0, 10)) {
        const versions = Object.keys(inc.versions).join(", ");
        printText(`  ${inc.package}: ${versions} → recommend ${inc.recommended}`);
      }
    }

    if (consolidationOpps.length > 0) {
      printText(`\nConsolidation opportunities:`);
      for (const opp of consolidationOpps) {
        printText(`  ${opp.category}: ${opp.packages.join(" + ")} → use ${opp.recommended}`);
        printText(`    ${opp.reason} (affects ${opp.projects.length} projects)`);
      }
    }

    if (inconsistencies.length === 0 && consolidationOpps.length === 0) {
      printText("Dependencies are well-standardized across projects.");
    }
  }
}
